modules/home-age.nix

```diff
@@ -0,0 +1,232 @@
+    )
+    chmod ${secretType.mode} "$TMP_FILE"
+    mv -f "$TMP_FILE" "$_truePath"
+
+    ${optionalString secretType.symlink ''
+      [ "${secretType.path}" != "${cfg.secretsDir}/${secretType.name}" ] && ln -sfn "${cfg.secretsDir}/${secretType.name}" "${secretType.path}"
+    ''}
+  '';
+
+  testIdentities =
+    map
+      (path: ''
+        test -f ${path} || echo '[agenix] WARNING: config.age.identityPaths entry ${path} not present!'
+      '')
+      cfg.identityPaths;
+
+  cleanupAndLink = ''
+    _agenix_generation="$(basename "$(readlink ${cfg.secretsDir})" || echo 0)"
+    (( ++_agenix_generation ))
+    echo "[agenix] symlinking new secrets to ${cfg.secretsDir} (generation $_agenix_generation)..."
+    ln -sfn "${cfg.secretsMountPoint}/$_agenix_generation" ${cfg.secretsDir}
+
+    (( _agenix_generation > 1 )) && {
+    echo "[agenix] removing old secrets (generation $(( _agenix_generation - 1 )))..."
+    rm -rf "${cfg.secretsMountPoint}/$(( _agenix_generation - 1 ))"
+    }
+  '';
+
+  installSecrets = builtins.concatStringsSep "\n" (
+    [ "echo '[agenix] decrypting secrets...'" ]
+    ++ testIdentities
+    ++ (map installSecret (builtins.attrValues cfg.secrets))
+    ++ [ cleanupAndLink ]
+  );
+
+  secretType = types.submodule ({ config, name, ... }: {
+    options = {
+      name = mkOption {
+        type = types.str;
+        default = name;
+        description = ''
+          Name of the file used in ''${cfg.secretsDir}
+        '';
+      };
+      file = mkOption {
+        type = types.path;
+        description = ''
+          Age file the secret is loaded from.
+        '';
+      };
+      path = mkOption {
+        type = types.str;
+        default = "${cfg.secretsDir}/${config.name}";
+        description = ''
+          Path where the decrypted secret is installed.
+        '';
+      };
+      mode = mkOption {
+        type = types.str;
+        default = "0400";
+        description = ''
+          Permissions mode of the decrypted secret in a format understood by chmod.
+        '';
+      };
+      symlink = mkEnableOption "symlinking secrets to their destination" // { default = true; };
+    };
+  });
+
+  mountingScript =
+    let
+      drv = pkgs.writeShellScript "agenix-home-mount-secrets" ''
+        set -e
+        set -o pipefail
+        export PATH="${pkgs.gnugrep}/bin:${pkgs.coreutils}/bin:@out@/sw/bin:/usr/bin:/bin:/usr/sbin:/sbin"
+        ${newGeneration}
+        ${installSecrets}
+        exit 0
+      '';
+    in
+    toString drv;
+
+  userDirectory = dir:
+    let
+      inherit (pkgs.stdenv.hostPlatform) isDarwin;
+      baseDir =
+        if isDarwin
+        then ''"$(getconf DARWIN_USER_TEMP_DIR)"''
+        else ''"$XDG_RUNTIME_DIR"'';
+    in
+    "${baseDir}/${dir}";
+
+  userDirectoryDescription = dir: ''
+    "$XDG_RUNTIME_DIR"/${dir} on linux or "$(getconf DARWIN_USER_TEMP_DIR)"/${dir} on darwin.
+  '';
```

Comment:
I just noticed that this means the `path` isn't really usable by other modules, since it relies on shell subtitution...

I wonder how `sops-nix` deals with this.

Reply:
It looks like they don't care, just using `%r` as a placeholder but not otherwise exposing the _actual_ path to be consumed by other modules.

Reply:
Looks like `homeage` also relies on an environment variable in its `path` option.

Reply:
I think I'm happy with that status quo, given that most usage of secrets are to expose environment variables through `home.sessionVariables` (which uses the shell) or for `password-command`-like options, which can also just use a `writeScript` wrapper to do the command substitution correctly.